pkg/processor/transform/txfbuiltin/httprequest.go

```diff
@@ -0,0 +1,181 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package txfbuiltin
+
+import (
+	"bytes"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/conduitio/conduit/pkg/foundation/cerrors"
+	"github.com/conduitio/conduit/pkg/processor"
+	"github.com/conduitio/conduit/pkg/processor/transform"
+	"github.com/conduitio/conduit/pkg/record"
+	"github.com/jpillora/backoff"
+)
+
+const (
+	httpRequestName = "httprequest"
+
+	httpRequestConfigURL          = "url"
+	httpRequestConfigMethod       = "method"
+	httpRequestBackoffRetryCount  = "backoffRetry.count"
+	httpRequestBackoffRetryMin    = "backoffRetry.min"
+	httpRequestBackoffRetryMax    = "backoffRetry.max"
+	httpRequestBackoffRetryFactor = "backoffRetry.factor"
+)
+
+func init() {
+	processor.GlobalBuilderRegistry.MustRegister(httpRequestName, transform.NewBuilder(HTTPRequest))
+}
+
+// HTTPRequest builds a transform that sends an HTTP request to the specified
+// URL with the specified HTTP method (default is POST). The record payload is
+// used as the request body and the raw response body is put into the record
+// payload.
+func HTTPRequest(config transform.Config) (transform.Transform, error) {
+	return httpRequest(httpRequestName, config)
+}
+
+func httpRequest(
+	transformName string,
+	config transform.Config,
+) (transform.Transform, error) {
+	var (
+		err    error
+		rawURL string
+		method string
+	)
+
+	if rawURL, err = getConfigFieldString(config, httpRequestConfigURL); err != nil {
+		return nil, cerrors.Errorf("%s: %w", transformName, err)
+	}
+
+	_, err = url.Parse(rawURL)
+	if err != nil {
+		return nil, cerrors.Errorf("%s: error trying to parse url: %w", transformName, err)
+	}
+
+	method = config[httpRequestConfigMethod]
+	if method == "" {
+		method = http.MethodPost
+	}
+
+	// preflight check
+	_, err = http.NewRequest(
+		method,
+		rawURL,
+		bytes.NewReader([]byte{}),
+	)
+	if err != nil {
+		return nil, cerrors.Errorf("%s: error trying to create HTTP request: %w", transformName, err)
+	}
+
+	txfFunc := func(r record.Record) (record.Record, error) {
+		req, err := http.NewRequest(
+			method,
+			rawURL,
+			bytes.NewReader(r.Payload.Bytes()),
+		)
+		if err != nil {
+			return record.Record{}, cerrors.Errorf("%s: error trying to create HTTP request: %w", transformName, err)
+		}
+
+		resp, err := http.DefaultClient.Do(req)
+		if err != nil {
+			return record.Record{}, cerrors.Errorf("%s: error trying to execute HTTP request: %w", transformName, err)
+		}
+		defer resp.Body.Close()
+
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return record.Record{}, cerrors.Errorf("%s: error trying to read response body: %w", transformName, err)
+		}
+
+		if resp.StatusCode > 299 {
```

Comment:
We can maybe reduce this to just 429 + some 5xx codes. 3xx will all be redirects, i.e. not errors, so there's probably no need to retry those. 4xx would be client errors, and they will mostly fail again (except for maybe 408 and 429). Similar for  the 5xx ones.

Reply:
I agree with you, those codes you mention probably don't make sense to be retried, although they also don't make sense to succeed IMO, at least it would be unexpected to me if they would mean the transform was successful. Given the current approach it wouldn't be so easy to skip those retries (the retry logic has no idea about the status code). Then again, even if we add this logic, there will inevitably be users that need more control, ideally I think we could have a list of successful and failed status codes that the user can configure (and sane defaults).

That said, I hope this solution is good enough for now, given that this processor was meant to address an immediate problem. So unless @lyuboxa says this doesn't cover their use case, I'd say we can expand this functionality in the future.

Reply:
I think we are good for now, we will iterate as the need arises.